{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Constructing phylogenetic trees using Maximum Likelihood\n",
    "This Jupyter notebook is created as a project in the Algorithms in Bioinformatics course 2018 given by [Swedish National Graduate School in Medical Bioinformatics](http://www.medbioinfo.se/).   \n",
    "In this notebook I have \n",
    "\n",
    "1. Multiple Sequence Alignment\n",
    "2. Generating a random Tree\n",
    "\n",
    "\n",
    "So lets start with setting up the working environment\n",
    "\n",
    "Tree topology\n",
    "Branch lengths\n",
    "Nucleotide frequencies\n",
    "Nucleotide-nucleotide substitution rates\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Maximum Likelihood"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiple Sequence Alignment\n",
    "The first step in creating a phylogenetic tree is to do a multiple sequence alignment (MSA) of the sequences that we want to include in our tree. This is done to find [homologous](https://en.wikipedia.org/wiki/Sequence_homology) posititions in the sequences, that is finding positions that have a shared ancestry. There are many  different tools for doing this, but the most common ones are [ClustalW](https://www.ebi.ac.uk/Tools/msa/clustalo/), [MUSCLE](https://www.ebi.ac.uk/Tools/msa/muscle/) and [MAFFT](https://www.ebi.ac.uk/Tools/msa/mafft/).\n",
    "An MSA can be represented in many different ways and there are many different file formats for storing MSAs. The easiest way to represnt an MSA is to each row represent a sequnce in the alignment and each column represent a position in the alignment. If the alignment was done on the protein level each cell will contain either an ammino acid or a deletion sign, usually a -. In python we can represent an MSA as a dictionary where each key is the sequence identifier and each value is a list containing the element at each position in the alignment.\n",
    "Next we will create a function that can read an MSA stored in a the plain text fasta-format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_msa(path):\n",
    "    \"\"\"\n",
    "    Read a multiple sequence alignment (MSA) in fasta format\n",
    "    and returns a dictionary where each key is the sequence\n",
    "    identifier and each value is a list with the elements\n",
    "    of the alignmnent.\n",
    "    \n",
    "    Args:\n",
    "    path (str) - Path to the fasta file.\n",
    "    \n",
    "    Return:\n",
    "    msa (dict) - The MSA in a dictionary\n",
    "    \"\"\"\n",
    "    msa = {}\n",
    "    with open(path, 'r') as fasta_file:\n",
    "        lines = fasta_file.readlines()\n",
    "        for i in range(len(lines)):\n",
    "            if lines[i][0] == '>':\n",
    "                seq_id = lines[i][1:].strip()\n",
    "                alignment = lines[i+1].strip()\n",
    "                msa[seq_id] = list(alignment)\n",
    "    \n",
    "    return msa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Seq_C': ['C'], 'Seq_E': ['C'], 'Seq_B': ['A'], 'Seq_A': ['A'], 'Seq_D': ['G']}\n"
     ]
    }
   ],
   "source": [
    "msa = read_msa('./data/msa_test.faa')\n",
    "print(msa)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The simple case\n",
    "We start with a simple case with an alignment of two sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_jc_distance(substitutions, identical):\n",
    "    \"\"\"\n",
    "    Calculate the distance between two sequences\n",
    "    \"\"\"\n",
    "    distance = -(3/4) * np.log(1-(4/3) * (substitutions/(substitution + identical)))\n",
    "    return distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_substitutions(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Function to count the number of susbtitutions in a pairwise alignmnent\n",
    "    \"\"\"\n",
    "    n_subst = 0\n",
    "    for index in range(len(seq1)):\n",
    "        if seq1[index] != seq2[index]:\n",
    "            n_subst += 1\n",
    "    return n_substitutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prob_no_mutation_jc(distance):\n",
    "    \"\"\"\n",
    "    Calculate the probability to not observe a mutation for a given distance using Jukes-Cantor model\n",
    "    \"\"\"\n",
    "    prob = (1/4) * (1 + 3*np.exp((-4/3) * distance))\n",
    "    return prob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prob_mutation_jc(distance):\n",
    "    \"\"\"\n",
    "    Calculate the probability to observe a mutation for a given distance using Jukes-Cantor model.\n",
    "    \"\"\"\n",
    "    prob = (1/4) * (1 - np.exp((-4/3) * distance))\n",
    "    return prob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tree structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Node:\n",
    "    def __init__(self, alignment_length, alignment=False, left=None, right=None, center=None):\n",
    "        self.left = left\n",
    "        self.right = right\n",
    "        self.center = center\n",
    "        self.length = 0.1\n",
    "        self.alignment_seq = alignment\n",
    "        self.alignment_length = alignment_length \n",
    "        self.nucl_map = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "        self.set_nucl_probs()\n",
    "        \n",
    "        \n",
    "    def set_nucl_probs(self):   \n",
    "        if self.alignment_seq:\n",
    "            self.nucl_prob = np.zeros((4, self.alignment_length)) \n",
    "            for index, nucleotide in enumerate(self.alignment_seq):\n",
    "                row = self.nucl_map[nucleotide]\n",
    "                self.nucl_prob[row, index] = 1  # observed nucleotide\n",
    "        else:\n",
    "            self.node_prob()\n",
    "    \n",
    "    def node_prob(self):\n",
    "        # This function should be moved to the Tree class\n",
    "        self.nucl_prob = np.zeros((4, self.alignment_length))\n",
    "        for index in range(self.alignment_length):\n",
    "            for s, nucl_node in enumerate(['A', 'C', 'G', 'T']):\n",
    "                if self.center:  # only used for root node\n",
    "                    center_prob = 0\n",
    "                right_prob = 0\n",
    "                left_prob = 0\n",
    "                for x, nucl_children in enumerate(['A', 'C', 'G', 'T']):\n",
    "                    if nucl_node == nucl_children:\n",
    "                        right_prob += prob_no_mutation_jc(self.right.length) * self.right.nucl_prob[x, index]\n",
    "                        left_prob += prob_no_mutation_jc(self.left.length) * self.left.nucl_prob[x, index]\n",
    "                        if self.center:\n",
    "                            center_prob += prob_no_mutation_jc(self.center.length) * self.center.nucl_prob[x, index]\n",
    "                    else:\n",
    "                        right_prob += prob_mutation_jc(self.right.length) * self.right.nucl_prob[x, index]\n",
    "                        left_prob += prob_mutation_jc(self.left.length) * self.left.nucl_prob[x, index]\n",
    "                        if self.center:\n",
    "                            center_prob += prob_mutation_jc(self.center_distance) * self.center.nucl_prob[x, index] \n",
    "                if self.center:                 \n",
    "                    self.nucl_prob[s, index] = np.sum(right_prob) * np.sum(left_prob) * np.sum(center_prob)\n",
    "                else:\n",
    "                    self.nucl_prob[s, index] = np.sum(right_prob) * np.sum(left_prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [],
   "source": [
    "# http://www.lacim.uqam.ca/~chauve/Enseignement/BIF7001/H04/Documents/Settles-ML-Branches.pdf\n",
    "# Must find how the first and second derivatives are calculated to finish this function.\n",
    "def optimize_branch_length(node, tree):\n",
    "    epsilon = 0.001\n",
    "    original_length = length\n",
    "    old_likelihood = 0\n",
    "    done = False\n",
    "    better = False\n",
    "    it = 1\n",
    "    ite = 0\n",
    "    \n",
    "    while it <4 and ite < 20 and not done:\n",
    "        likelihood = compute_likelihood(tree)\n",
    "        if likelihood > old_likelihood:\n",
    "            old_length = y\n",
    "            old_likelihood = likelihood\n",
    "            better = True\n",
    "            it += 1\n",
    "        if better:\n",
    "            node.length = node.length + first_derative/abs(second_derative)\n",
    "        else:\n",
    "            node.length = (node.length + 19 * yold)/20\n",
    "        ite += 1\n",
    "        if abs(node.length-old_length) < epsilon:\n",
    "            done = True\n",
    "    p_length = yold\n",
    "    tree.log_likelihood = old_likelihood\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Tree:\n",
    "    def __init__(self, aln):\n",
    "        self.alignment = aln\n",
    "        self.set_alignment_length()\n",
    "        self.create_tree()\n",
    "        self.log_likelihood = self.get_prob()\n",
    "        \n",
    "    def set_alignment_length(self):\n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        for aln in self.alignment.keys():\n",
    "            self.alignment_length = len(self.alignment[aln])\n",
    "            break\n",
    "        \n",
    "    def create_tree(self):\n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        nodes = []\n",
    "        for seq in self.alignment.keys():\n",
    "            node = Node(alignment_length=self.alignment_length, alignment=self.alignment[seq])\n",
    "            nodes.append(node)\n",
    "        \n",
    "        while len(nodes) > 3:\n",
    "            node = Node(alignment_length=self.alignment_length, left=nodes[0], right=nodes[1])\n",
    "            nodes.pop(0)\n",
    "            nodes.pop(0)\n",
    "            nodes.append(node)\n",
    "        \n",
    "        self.root = Node(alignment_length=self.alignment_length, left=nodes[0], right=nodes[1], center=nodes[2])\n",
    "        self.root.length = None\n",
    "    \n",
    "    def calculate_log_likelihood(self):\n",
    "        # TODO: Instead of calculating the log-likelihood when the tree topology,\n",
    "        # implement the Felsensteins pruning algorithm here. Then the log-likelihood\n",
    "        # can be updated when there are changes in the branch lengths.\n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        print(np.sum(0.25 * self.root.nucl_prob.sum(axis=0)))\n",
    "        log_likelihoods = np.log(0.25 * self.root.nucl_prob.sum(axis=0))\n",
    "        return np.sum(log_likelihoods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Implement a tree-traversal to visit all nodes to optimize the branch-lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.10391849419151006\n",
      "-27.943487826273014\n",
      "0.10366980692593511\n",
      "-29.000039572820285\n"
     ]
    }
   ],
   "source": [
    "aln = {'seq1':'CCAG', 'seq2':'CCTG', 'seq3':'CCAG', 'seq4':'GCAG', 'seq5':'ACAA', 'seq6':'CCTA'}\n",
    "tree = Tree(aln)\n",
    "print(tree.log_likelihood)\n",
    "aln = {'seq1':'ACAG', 'seq2':'CCTG', 'seq3':'CCAG', 'seq4':'GCAG', 'seq5':'CCAA', 'seq6':'CCTA'}\n",
    "tree = Tree(aln)\n",
    "print(tree.log_likelihood)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Felstein's pruning algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-5.205091174921215"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for nucl in ['A', 'C', 'G', 'T']\n",
    "node_prob[nucl] = node.right.prob(nucl) + node.left.prob(nucl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQUAAABqCAYAAAC1Q0SqAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAN1wAADdcBQiibeAAACtpJREFUeJzt3V1MVNcaBuB3YLBF/kotiqCO1IqFdCiNAQSj5aKtNVALTWmtTYmpqL0opTG9O03rmJjYJlhi1NMqtmCNps7Earxqk3McYpR2KqYyEjv8dZAzxlpEj8wwCmdY5wIZXY781Cp73Ot9Em/23mO+DfF17bW+tccghBAgIropQusCiCi8MBSISMJQ0LHa2loYDAacPHlS61LoIcJQ0DGr1YqVK1fCarVqXUrY+vzzzzF9+nT4/X6tSwkbDAWd6u3thcvlwpYtW3Do0CFwPhnY8s1/8GSRA9X13cFj9fX1KC0txYEDBzSsLLwYtS6AHozDhw+juLgYJpMJSUlJ+Pnnn7Fo0SLpGovFolF1k6/14jTYmrIxMDiEf+zoQuZTMZj6PydSUlJQUVGB9957D++++67WZYYFjhR0ymq14uWXX0Z/fz+KiopCHiEsFguOHTsGIYQSfwCBQCAQvH+DYXjOZe3atcjJycHQ0BB++umnSf4thScD+xT058qVKzCZTJg5cyYAQAiBgYEBuN3u4DUWiwVCCGzcuFGbIjWwda8HO6wefPBWKsqXT8WcOXOQk5ODqKgodHR0ID8/H99++63WZWqOjw86dOTIEbz55pvYvXt38Nizzz4Lh8OB3NxcDSvT1obyVGwoTwUAbN++HWvWrEFNTQ0AoK+vD2lpafjzzz+RlJSkZZma4+ODDlmtVpSVlUnHysrKYLVag5NtJ9tma1RdeNi9e7c0hxAXF4fXX38dtbW1GlYVHpR/fFB1si0yQmBlXjP27azUuiwKM0qPFFSebDMIAFD6/wMahfJzCoWFhUpNti28OdmWnngWT03v1bocCkNKjxRUtKE8FR1Hc7FoXvf4FyuAHY2hGAqkDHY0Tozyjw8qa2hoUGailR2NE8dQUNSnn3562wSkCkYmWQ0ARu9ovLMVXEW6D4WRX/yJEydQUFCgdTlhRaUJVuDWJOsHb6UiL8OAssOH4fF4sGfPHvT19WHHjh0MBSjQp7Bs2TI8/vjjSE5OxhdffCGdU7HVl4Zt374d7e3tIR2N586dY0ej1gU8SNw+TKNhR+PodP34MJHtw6SmM2fOhBz78ssvNagk/Oh6pDDe9mEiCqXbkcKVK1dw4sQJdHZ2Ari1fbi6ulrjyojCm25HCiPbh10uF1wuF1pbW5GQkACHw6F1aURhTbehMOb24TpuH9Y7vsl6Ynw+n/RGKuDmkqQqXW0Atw+rYqylaAK6urqwatUqeL1eDA4O4qWXXgouz0YA0HxLL7cP0/3EpehQt+/78Pv9KCoqwsaNG3HmzBk4nU64XC4cPHgQwM2JRtWad7h9WN/4JmvZnfs+Bv97GgsWLMCLL74IAIiMjERdXR2uX78OQMdzCmPh9mF945usRx8dA0BnuxMZGRnSz2TGjBkwmUwAdLwkSWqa6FK0qi/X+eCtVKTF9gYfFUZcunQJ3d3dWLhwoZojBdIvLkXf3cjouGpVKpYsWYKmpiY4nU4Aw8G5YcMGnDp1CgBHCqQzVqsVVVVV0rGRpeh/t6Rgl+0C5ifORv5T5zWqUHuJiYnYv38/KisrERkZiatXr2Lp0qVYt24dAIaCro23bVyPk225ublobGxEY2OjdPyC14Rt/3RjYHAI5y/OQ1K8T6MKw8Nzzz0Hu90On8+H6OhoRETcemjg44OOjfWt0ypPtnEp+paYmBgpEACOFHRrZK2+oaEBS5cuxdatW2EwGKRrVJ1s41L0LT6fD48++igiIyODxzhS0Km7rdWrjkvRt3R1dWHx4sUoKCiA2WzGhx9+GDzHUNApbhun2/3ljkbSF24bH5/Kb7Ier6ORoaBD/Nbpsan+JuvROhpHMBR0iGv141NpghX4ax2Nun+b81gsFgvsdjsKCwu1LmVScNs4AcOPlwUFBTh48CDMZjOEEHjnnXewZMkSrF+/Xu2RgsrDSK7Vq2u8jkalRwoq2nrbWn3ek+eVG0aTjB2NxLV6ktyto5GhQEQShgIpb3BwENeuXdO6jLCh9EQjqe3y5ct4++230dPTE9wXsnfv3pA1fNVwpEDKuL3VNxAI4LXXXsOaNWtw6tQp/PLLL9i0aRMqKiq0LlNzHCkoTOVW3ylDHTAajdJ3gyxfvhzx8fEaVhkeGAqKUrlHAwDazp2G2WwOnq2trcW1a9cghEBmZiYSExM1qlN7DAWFqdajcHurrzl5ANXV/wqeS05ORlxcHD755BMUFxcrHQpsXiIl+f1+5OXlYfPmzXjllVcADE885ufn4+jRo1iwYIHGFWqHoUDK+uOPP7Bu3Tp4PB4YjUZ4vV5UVlaioqJCehORahgKpLxAIACv14uEhAStSwkLBrvdLoQQD/1OwYaGBujhPoi0FmG322G327Wu42/Ty30QaY3NS0QkYSgQkYShQEQSo9vthtvtfujbXe12O+bOnat1GUQPPSMA3bS76uEeiLRmnDt3Lkwm00Pf8mqxWBgKRPcB5xSISMJQICIJQ4GIJMbCwkLpWfzGjRt45JFHJvRhv9+P6Ojou57r7+/H1KlT70uRE3HnfRDRvYl4/vnnUVhYiNOnT2Px4sWIj49Heno6fvjhh1E/tGnTJsyfPx9JSUkoKSnB2bNng+e+/vprzJgxAwkJCcjLy8O+ffsm4z4wch9E9DcJIYTf7xcZGRnis88+EwMDA+L7778XsbGx4vz58+JOx48fF8nJyaKlpUX09vaK1atXi6KiIiGEEL///ruIjo4W27ZtE319faK+vl4kJiYKn88X8vcQUXiKAIDjx4+jp6cHH330EaKiolBSUoKUlBRYrdaQEDl27BiysrKCr6xasWIFmpqaAAC//vorli1bhsrKSsTGxqK0tBR+vx9Op3Nyk46I7pkRAFwuF7Kzs6VviklPT0dra2vIB9544w1UV1fj448/xuzZs1FTU4O1a9cCAEpKSlBSUoL+/n40Nzfjq6++gslkQk5OziTdDhH9XUYA6OzsxGOPPSadMJlMuHr1asgHpk2bhqysLOzcuROpqalob29HWlqadI3b7UZ5eTna2tqwevVqXL9+fVInHYno3kUAw98nd+nSJemE1+vF008/HfKBqqoqREVF4eLFi3A6nThw4ADef/99eL3e4DWZmZlobW2Fx+PBoUOHsGfPngd8G0R0v0QAwLx589DR0SGdaGtrQ1ZWVsgHHA4HVqxYgSlTpgAAXn31VQQCATQ2NqKmpgabN28OXpuSkoIXXngBzc3ND/IeiOg+igCG/2H39PTAZrNBCIH9+/fD6XQiNzcXAPDdd9+hpaUFAGA2m2Gz2XD58mUAgM1mAwA888wzyM7Oxq5du+BwOAAAv/32G3788Ufk5+dP+o0R0T0aWYY4cuSIiI6OFjNnzhSxsbHCZrMFlyhmzZolampqhBBCdHd3i+LiYhEXFyfmzJkjYmJiRF1dnRBCiEAgIEpLSwUAMWvWLGEwGMT69evF0NDQ5K6pENE9k97mfOPGDbhcLmRkZCAqKmrMMOnr64PL5UJ6enrIV215PB5cuHABaWlpeOKJJx5ImBHRg8FXvBORhBuiiEjCUCAiCUOBiCQMBSKSMBSISMJQICIJQ4GIJAwFIpIwFIhIwlAgIglDgYgkDAUikjAUiEjCUCAiCUOBiCQMBSKSMBSISMJQICIJQ4GIJAwFIpIwFIhIwlAgIglDgYgkDAUikjAUiEjCUCAiCUOBiCQMBSKS/B8ghjA3wValCAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t.render('%%inline')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
